File: lib/nc/core/sync.ex
defmodule Nc.Core.Sync do
  # similar to the `xform` function in the jupiter OT model, but working with blocks of text instead of individual text operations

  # Excerpt from the paper:
  # "The general tool for handling conflicting messages is a function, xform,
  # that maps a pair of messages to the fixed up versions. We write
  # Xform(c, s) = { c', s' }
  # where c and s are the original client and server messages.
  # The messages c' and s' must have the property that if the client applies c
  # followed by s', and the server applies s followed by c', then the client and
  # server will wind up in the same final state."

  @type change() ::
          {
            :insert,
            position :: non_neg_integer(),
            text :: String.t(),
            from :: any()
          }
          | {
              :delete,
              position :: non_neg_integer(),
              amount :: non_neg_integer(),
              from :: any()
            }
          | nil

  @spec reconcile(change(), change()) :: {change(), change()}

  # this function must be commutative!
  def reconcile({:insert, position_1, text_1, from_1}, {:insert, position_2, text_2, from_2}) do
    cond do
      position_1 > position_2 || (position_1 == position_2 && text_1 > text_2) ||
          (position_1 == position_2 && text_1 == text_2 && from_1 > from_2) ->
        {
          {:insert, position_1 + String.length(text_2), text_1, from_1},
          {:insert, position_2, text_2, from_2}
        }

      position_1 < position_2 || (position_1 == position_2 && text_1 < text_2) ||
          (position_1 == position_2 && text_1 == text_2 && from_1 < from_2) ->
        {
          {:insert, position_1, text_1, from_1},
          {:insert, position_2 + String.length(text_1), text_2, from_2}
        }

      # the case for identical inputs
      # this is actually has the benefit of cancelling out changes coming from the same source
      true ->
        {nil, nil}
    end
  end

  # for simplicity's sake, if an insert is inside a delete, it is nullified
  def reconcile({:insert, position_1, text_1, from_1}, {:delete, position_2, amount_2, from_2}) do
    if position_1 > position_2 do
      if position_1 >= position_2 + amount_2 do
        {
          {:insert, position_1 - amount_2, text_1, from_1},
          {:delete, position_2, amount_2, from_2}
        }
      else
        {
          nil,
          {:delete, position_2, amount_2 + String.length(text_1), from_2}
        }
      end
    else
      {
        {:insert, position_1, text_1, from_1},
        {:delete, position_2 + String.length(text_1), amount_2, from_2}
      }
    end
  end

  def reconcile({:delete, position_1, amount_1, from_1}, {:insert, position_2, text_2, from_2}) do
    {insert_prime, delete_prime} =
      reconcile({:insert, position_2, text_2, from_2}, {:delete, position_1, amount_1, from_1})

    {delete_prime, insert_prime}
  end

  def reconcile({:delete, position_1, amount_1, from_1}, {:delete, position_2, amount_2, from_2}) do
    left_1 = position_1
    right_1 = position_1 + amount_1
    left_2 = position_2
    right_2 = position_2 + amount_2

    overlap_start = max(left_1, left_2)
    overlap_end = min(right_1, right_2)
    overlap_area = overlap_end - overlap_start

    cond do
      overlap_area <= 0 ->
        if position_1 > position_2 do
          {
            {:delete, position_1 - amount_2, amount_1, from_1},
            {:delete, position_2, amount_2, from_2}
          }
        else
          {
            {:delete, position_1, amount_1, from_1},
            {:delete, position_2 - amount_1, amount_2, from_2}
          }
        end

      # left "eclipses" right
      left_1 <= left_2 && right_1 >= right_2 ->
        {
          {:delete, position_1, amount_1 - amount_2, from_1},
          nil
        }

      # right "eclipses" left
      left_1 >= left_2 && right_1 <= right_2 ->
        {
          nil,
          {:delete, position_2, amount_2 - amount_1, from_2}
        }

      # overlap on one side
      left_1 > left_2 || right_1 < right_2 ->
        {
          {:delete, min(position_2, position_1), amount_1 - overlap_area, from_1},
          {:delete, min(position_2, position_1), amount_2 - overlap_area, from_2}
        }
    end
  end

  def reconcile(nil, change) do
    {
      nil,
      change
    }
  end

  def reconcile(change, nil) do
    {
      change,
      nil
    }
  end

  @spec clamp(change(), String.t()) :: change()
  def clamp(change, string) do
    string_length = String.length(string)

    case change do
      {:delete, position, amount, from} ->
        position = position |> min(string_length) |> max(0)
        amount = amount |> min(string_length - position) |> max(0)

        if amount <= 0 do
          nil
        else
          {:delete, position, amount, from}
        end

      {:insert, position, text, from} ->
        position = position |> min(string_length) |> max(0)
        {:insert, position, text, from}
    end
  end

  def reconcile_against(incoming_list, outgoing_list) do
    reconcile_single_change = fn outgoing_change, {incoming_change, new_outgoing_list} ->
      {new_outgoing_change, new_incoming_change} =
        reconcile(outgoing_change, incoming_change)

      {new_incoming_change, [new_outgoing_change | new_outgoing_list]}
    end

    reconcile_all_changes = fn incoming_change, {new_incoming_list, current_outgoing_list} ->
      {new_incoming_change, new_outgoing_list} =
        Enum.reduce(current_outgoing_list, {incoming_change, []}, reconcile_single_change)

      new_outgoing_list = Enum.reverse(new_outgoing_list)
      {[new_incoming_change | new_incoming_list], new_outgoing_list}
    end

    {new_incoming_list, new_outgoing_list} =
      Enum.reduce(incoming_list, {[], outgoing_list}, reconcile_all_changes)

    {Enum.reverse(new_incoming_list), new_outgoing_list}
  end
end

File: client/src/lib/sync.ts
import type { Change, Insert, Delete } from "./messenger";

// this is translated directly from my implementation in elixir

function reconcile(
  change1: Change,
  change2: Change
): { change1: Change; change2: Change } {
  if (!change1 || !change2) {
    return {
      change1,
      change2,
    };
  } else {
    if (change1.type == "insert" && change2.type == "insert") {
      change1 = change1 as Insert;
      change2 = change2 as Insert;
      return insert_insert(change1, change2);
    } else if (change1.type == "insert" && change2.type == "delete") {
      change1 = change1 as Insert;
      change2 = change2 as Delete;
      return insert_delete(change1, change2);
    } else if (change1.type == "delete" && change2.type == "insert") {
      change1 = change1 as Delete;
      change2 = change2 as Insert;
      let rec = insert_delete(change2, change1);
      return {
        change1: rec.change2,
        change2: rec.change1,
      };
    } else {
      change1 = change1 as Delete;
      change2 = change2 as Delete;
      return delete_delete(change1, change2);
    }
  }
}

function insert_insert(
  change1: Insert,
  change2: Insert
): { change1: Change; change2: Change } {
  if (
    change1.position > change2.position ||
    (change1.position == change2.position && change1.text > change2.text) ||
    (change1.position == change2.position &&
      change1.text == change2.text &&
      change1.from > change2.from)
  ) {
    return {
      change1: {
        ...change1,
        position: change1.position + change2.text.length,
      },
      change2,
    };
  } else if (
    change1.position < change2.position ||
    (change1.position == change2.position && change1.text < change2.text) ||
    (change1.position == change2.position &&
      change1.text == change2.text &&
      change1.from < change2.from)
  ) {
    return {
      change1,
      change2: {
        ...change2,
        position: change2.position + change1.text.length,
      },
    };
  } else {
    return { change1: undefined, change2: undefined };
  }
}

function insert_delete(
  change1: Insert,
  change2: Delete
): { change1: Change; change2: Change } {
  if (change1.position > change2.position) {
    if (change1.position >= change2.position + change2.amount) {
      return {
        change1: {
          ...change1,
          position: change1.position - change2.amount,
        },
        change2,
      };
    } else {
      return {
        change1: undefined,
        change2: {
          ...change2,
          amount: change2.amount + change1.text.length,
        },
      };
    }
  } else {
    return {
      change1,
      change2: {
        ...change2,
        position: change2.position + change1.text.length,
      },
    };
  }
}

function delete_delete(
  change1: Delete,
  change2: Delete
): { change1: Change; change2: Change } {
  const left_1 = change1.position;
  const right_1 = change1.position + change1.amount;
  const left_2 = change2.position;
  const right_2 = change2.position + change2.amount;

  const overlap_start = Math.max(left_1, left_2);
  const overlap_end = Math.min(right_1, right_2);
  const overlap_area = overlap_end - overlap_start;

  if (overlap_area <= 0) {
    if (change1.position > change2.position) {
      return {
        change1: {
          ...change1,
          position: change1.position - change2.amount,
        },
        change2,
      };
    } else {
      return {
        change1,
        change2: {
          ...change2,
          position: change2.position - change1.amount,
        },
      };
    }
  } else if (left_1 <= left_2 && right_1 >= right_2) {
    return {
      change1: {
        ...change1,
        amount: change1.amount - change2.amount,
      },
      change2: undefined,
    };
  } else if (left_1 >= left_2 && right_1 <= right_2) {
    return {
      change1: undefined,
      change2: {
        ...change2,
        amount: change2.amount - change1.amount,
      },
    };
  } else {
    return {
      change1: {
        ...change1,
        position: Math.min(change2.position, change1.position),
        amount: change1.amount - overlap_area,
      },
      change2: {
        ...change2,
        position: Math.min(change2.position, change1.position),
        amount: change2.amount - overlap_area,
      },
    };
  }
}

// why doesnt javascript have tuples???
export function reconcileAgainst(
  incoming_list: Change[],
  outgoing_list: Change[]
): { new_incoming_list: Change[]; new_outgoing_list: Change[] } {
  const { new_incoming_list, new_outgoing_list } = incoming_list.reduce(
    reconcile_all_changes,
    { new_incoming_list: [], new_outgoing_list: outgoing_list }
  );

  return { new_incoming_list: new_incoming_list.reverse(), new_outgoing_list };
}

function reconcile_all_changes(
  changeObj: {
    new_incoming_list: Change[];
    new_outgoing_list: Change[];
  },
  incoming_change: Change
): {
  new_incoming_list: Change[];
  new_outgoing_list: Change[];
} {
  const new_incoming_list = changeObj.new_incoming_list;
  const current_outgoing_list = changeObj.new_outgoing_list;

  let { new_incoming_change, new_outgoing_list } = current_outgoing_list.reduce(
    reconcile_single_change,
    {
      new_incoming_change: incoming_change,
      new_outgoing_list: [],
    }
  );

  new_outgoing_list = new_outgoing_list.reverse();

  return {
    new_incoming_list: [new_incoming_change, ...new_incoming_list],
    new_outgoing_list: new_outgoing_list,
  };
}

function reconcile_single_change(
  changeObj: {
    new_incoming_change: Change;
    new_outgoing_list: Change[];
  },
  outgoing_change: Change
): {
  new_incoming_change: Change;
  new_outgoing_list: Change[];
} {
  const incoming_change = changeObj.new_incoming_change;
  const new_outgoing_list = changeObj.new_outgoing_list;

  const changeObj2 = reconcile(outgoing_change, incoming_change);

  const new_outgoing_change = changeObj2.change1;
  const new_incoming_change = changeObj2.change2;

  return {
    new_incoming_change,
    new_outgoing_list: [new_outgoing_change, ...new_outgoing_list],
  };
}
